### PR TITLE
refactor(python): much-improved `Instruction` matching for `BytecodeParser`

### PR DESCRIPTION
**Pull request by alexander-beedie**
**TLDR**:
* Cleans up the instruction-matching code; will make it easier to integrate more sophisticated matching later.
---

**Explanation**:

This puts matching specific `Instruction` sequences within the `RewrittenInstructions` class on a stronger footing; streamlines and consolidates the matching logic in a new method and allows for a much cleaner and less verbose (and less potentially error-prone) expression of matching rules.

For example, the following match definition finds (and returns) all `Instruction` sequences representing supported `numpy` function or `json.loads` calls as we advance through the operation stack:

**Before:**
```python
if inst.opname == "LOAD_GLOBAL" and (
    inst.argval in _NUMPY_MODULE_ALIASES or inst.argval == "json"
):
    instruction_buffer = list(islice(self._instructions, 3))
    if (
        len(instruction_buffer) == 3
        and (
            instruction_buffer[0].argval == "loads"
            or instruction_buffer[0].argval in _NUMPY_FUNCTIONS
        )
        and instruction_buffer[1].opname.startswith("LOAD_")
        and instruction_buffer[2].opname.startswith("CALL")
    ):
```
**After:**
```python
if matching_instructions := self._matches(
    idx,
    opnames=["LOAD_GLOBAL", "LOAD_*", "LOAD_*", "CALL*"],
    argvals=[_NUMPY_MODULE_ALIASES | {"json"}, _NUMPY_FUNCTIONS | {"loads"}],
):
```

### CODE REVIEW

File: py-polars/polars/utils/udfs.py [hunk 1]
```diff
@@ -392,19 +395,58 @@ def __iter__(self) -> Iterator[Instruction]:
...
+        self._instructions = list(instructions)
         updated_instructions: list[Instruction] = []
-        for inst in instructions:
-            if not any(
-                apply_rewrite(inst, updated_instructions)
+        idx = 0
+        while idx < len(self._instructions):
+            inst, increment = self._instructions[idx], 1
+            if inst.opname.startswith("LOAD_") and inst.opname not in _LOAD_OPS:
+                x = 1
```
Comment (MarcoGorelli) [inline py-polars/polars/utils/udfs.py:447]: what's x?

Comment (alexander-beedie) [reply py-polars/polars/utils/udfs.py]: Me forgetting to take out a debugging line - fixed! 🤣

**Comment by ritchie46** [comment]
@MarcoGorelli can you give the green light on this one?

**Review by MarcoGorelli — APPROVED** [review]
Nice work, this does simplify reading the code a bit
